File: mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxLifecycleAwareObserver.java
package com.airbnb.mvrx;

import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.Lifecycle.Event;
import android.arch.lifecycle.Lifecycle.State;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.LifecycleOwner;
import android.arch.lifecycle.OnLifecycleEvent;
import android.support.annotation.Nullable;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.internal.disposables.DisposableHelper;
import io.reactivex.internal.functions.Functions;
import io.reactivex.internal.observers.LambdaObserver;

@SuppressWarnings("unused")
public final class MvRxLifecycleAwareObserver<T> extends AtomicReference<Disposable> implements LifecycleObserver, Observer<T>, Disposable {

    public static final State DEFAULT_ACTIVE_STATE = State.STARTED;

    private final Observer<T> sourceObserver;
    private LifecycleOwner owner;

    @Nullable private Disposable sourceDisposable;
    @Nullable private T lastUndeliveredValue;
    @Nullable private T lastValue;
    private final AtomicBoolean locked = new AtomicBoolean(true);
    private final State activeState;
    private final boolean alwaysDeliverLastValueWhenUnlocked;

    public static class Builder<T> {
        private Consumer<? super T> onNext = Functions.emptyConsumer();
        private Consumer<? super Throwable> onError = Functions.ON_ERROR_MISSING;
        private Action onComplete = Functions.EMPTY_ACTION;
        private Consumer<? super Disposable> onSubscribe = Functions.emptyConsumer();
        private final LifecycleOwner owner;
        private Lifecycle.State activeState = DEFAULT_ACTIVE_STATE;
        private boolean alwaysDeliverValueWhenUnlocked = false;

        public Builder(LifecycleOwner owner) {
            this.owner = owner;
        }

        public Builder(LifecycleOwner owner, Lifecycle.State activeState) {
            this.owner = owner;
            this.activeState = activeState;
        }

        public Builder<T> alwaysDeliverValueWhenUnlocked() {
            return alwaysDeliverValueWhenUnlocked(true);
        }

        public Builder<T> alwaysDeliverValueWhenUnlocked(boolean shouldDeliver) {
            alwaysDeliverValueWhenUnlocked = shouldDeliver;
            return this;
        }

        public Builder<T> onNext(Consumer<? super T> onNext) {
            this.onNext = onNext;
            return this;
        }

        public Builder<T> onError(Consumer<? super Throwable> onError) {
            this.onError = onError;
            return this;
        }

        public Builder<T> onSubscribe(Consumer<? super Disposable> onSubscribe) {
            this.onSubscribe = onSubscribe;
            return this;
        }

        public Builder<T> onComplete(Action onComplete) {
            this.onComplete = onComplete;
            return this;
        }

        public MvRxLifecycleAwareObserver<T> build() {
            return new MvRxLifecycleAwareObserver<>(
                    owner,
                    activeState,
                    new LambdaObserver<T>(onNext, onError, onComplete, onSubscribe),
                    alwaysDeliverValueWhenUnlocked
            );
        }
    }

    public static <T> Builder<T> builder(LifecycleOwner owner) {
        return new Builder<>(owner);
    }

    public static <T> Builder<T> builder(LifecycleOwner owner, Lifecycle.State activeState) {
        return new Builder<>(owner, activeState);
    }

    public static <T> MvRxLifecycleAwareObserver<T> from(LifecycleOwner owner, Observer<T> consumer) {
        return new MvRxLifecycleAwareObserver<>(owner, consumer);
    }

    /**
     * Create a LifecycleObserver from a {@link Consumer}.
     */
    public static <T> MvRxLifecycleAwareObserver<T> from(LifecycleOwner owner, Consumer<T> consumer) {
        return MvRxLifecycleAwareObserver.<T>builder(owner).onNext(consumer).build();
    }

    public MvRxLifecycleAwareObserver(LifecycleOwner owner, Observer<T> consumer) {
        this(owner, DEFAULT_ACTIVE_STATE, consumer, false);
    }

    public MvRxLifecycleAwareObserver(
            LifecycleOwner owner,
            Lifecycle.State activeState,
            Observer<T> sourceObserver,
            boolean alwaysDeliverLastValueWhenUnlocked) {
        super();
        this.alwaysDeliverLastValueWhenUnlocked = alwaysDeliverLastValueWhenUnlocked;
        this.sourceObserver = sourceObserver;
        this.owner = owner;
        this.activeState = activeState;
    }

    @Override
    public void onSubscribe(Disposable d) {
        if (DisposableHelper.setOnce(this, d)) {
            owner.getLifecycle().addObserver(this);
            sourceObserver.onSubscribe(this);
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    void onDestroy() {
        owner.getLifecycle().removeObserver(this);
        owner = null;
        if (!isDisposed()) {
            dispose();
        }
    }

    @OnLifecycleEvent(Event.ON_ANY)
    void onLifecycleEvent() {
        updateLock();
    }

    private void updateLock() {
        if (owner != null && owner.getLifecycle().getCurrentState().isAtLeast(activeState)) {
            unlock();
        } else {
            lock();
        }
    }

    private void unlock() {
        if (!locked.getAndSet(false)) {
            return;
        }
        if (!isDisposed() ) {
            if (alwaysDeliverLastValueWhenUnlocked && lastValue != null) {
                sourceObserver.onNext(lastValue);
                lastUndeliveredValue = null;
            } else if (lastUndeliveredValue != null) {
                sourceObserver.onNext(lastUndeliveredValue);
                lastUndeliveredValue = null;
            }
        }
    }

    private void lock() {
        locked.set(true);
    }

    @Override
    public void onNext(T t) {
        if (!locked.get()) {
            sourceObserver.onNext(t);
        } else {
            lastUndeliveredValue = t;
        }
        lastValue = t;
    }

    @Override
    public void onError(Throwable e) {
        if (!isDisposed()) {
            lazySet(DisposableHelper.DISPOSED);
            sourceObserver.onError(e);
        }
    }

    @Override
    public void onComplete() {
        sourceObserver.onComplete();
    }

    @Override
    public void dispose() {
        DisposableHelper.dispose(this);
    }

    @Override
    public boolean isDisposed() {
        return get() == DisposableHelper.DISPOSED;
    }
}

File: mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxLifecycleAwareObserver.kt
package com.airbnb.mvrx

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.Lifecycle.Event
import androidx.lifecycle.Lifecycle.State
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.OnLifecycleEvent
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer
import io.reactivex.internal.disposables.DisposableHelper
import io.reactivex.internal.functions.Functions
import io.reactivex.internal.observers.LambdaObserver
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

/**
 * An wrapper around an [Observer] associated with a [LifecycleOwner]. It has an [activeState], and when in a lifecycle state greater
 * than the [activeState] (as defined by [Lifecycle.State.isAtLeast()]) it will deliver values to the [sourceObserver] or [onNext] lambda.
 * When in a lower lifecycle state, the most recent update will be saved, and delivered when active again.
 */
internal class MvRxLifecycleAwareObserver<T : Any>(
    private var owner: LifecycleOwner?,
    private val activeState: Lifecycle.State = DEFAULT_ACTIVE_STATE,
    private val deliveryMode: DeliveryMode = RedeliverOnStart,
    private var lastDeliveredValueFromPriorObserver: T?,
    private var sourceObserver: Observer<T>?,
    private val onDispose: () -> Unit
) : AtomicReference<Disposable>(), LifecycleObserver, Observer<T>, Disposable {

    constructor(
        owner: LifecycleOwner,
        activeState: Lifecycle.State = DEFAULT_ACTIVE_STATE,
        deliveryMode: DeliveryMode = RedeliverOnStart,
        lastDeliveredValue: T? = null,
        onComplete: Action = Functions.EMPTY_ACTION,
        onSubscribe: Consumer<in Disposable> = Functions.emptyConsumer(),
        onError: Consumer<in Throwable> = Functions.ON_ERROR_MISSING,
        onNext: Consumer<T> = Functions.emptyConsumer(),
        onDispose: () -> Unit
    ) : this(owner, activeState, deliveryMode, lastDeliveredValue, LambdaObserver<T>(onNext, onError, onComplete, onSubscribe), onDispose)

    private var lastUndeliveredValue: T? = null
    private var lastValue: T? = null
    private val locked = AtomicBoolean(true)
    private val isUnlocked
        get() = !locked.get()

    override fun onSubscribe(d: Disposable) {
        if (DisposableHelper.setOnce(this, d)) {
            requireOwner().lifecycle.addObserver(this)
            requireSourceObserver().onSubscribe(this)
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        requireOwner().lifecycle.removeObserver(this)
        if (!isDisposed) {
            dispose()
        }
        owner = null
        sourceObserver = null
    }

    @OnLifecycleEvent(Event.ON_ANY)
    fun onLifecycleEvent() {
        updateLock()
    }

    private fun updateLock() {
        if (owner?.lifecycle?.currentState?.isAtLeast(activeState) == true) {
            unlock()
        } else {
            lock()
        }
    }

    override fun onNext(nextValue: T) {
        if (isUnlocked) {
            val suppressRepeatedFirstValue = deliveryMode is UniqueOnly && lastDeliveredValueFromPriorObserver == nextValue
            lastDeliveredValueFromPriorObserver = null
            if (!suppressRepeatedFirstValue) {
                requireSourceObserver().onNext(nextValue)
            }
        } else {
            lastUndeliveredValue = nextValue
        }
        lastValue = nextValue
    }

    override fun onError(e: Throwable) {
        if (!isDisposed) {
            lazySet(DisposableHelper.DISPOSED)
            requireSourceObserver().onError(e)
        }
    }

    override fun onComplete() {
        requireSourceObserver().onComplete()
    }

    override fun dispose() {
        DisposableHelper.dispose(this)
        onDispose()
    }

    override fun isDisposed(): Boolean {
        return get() === DisposableHelper.DISPOSED
    }

    private fun unlock() {
        if (!locked.getAndSet(false)) {
            return
        }
        if (!isDisposed) {
            val valueToDeliverOnUnlock = when {
                deliveryMode is UniqueOnly -> lastUndeliveredValue
                deliveryMode is RedeliverOnStart && lastUndeliveredValue != null -> lastUndeliveredValue
                deliveryMode is RedeliverOnStart && lastUndeliveredValue == null -> lastValue
                else -> throw IllegalStateException("Value to deliver on unlock should be exhaustive.")
            }
            lastUndeliveredValue = null
            if (valueToDeliverOnUnlock != null) {
                onNext(valueToDeliverOnUnlock)
            }
        }
    }

    private fun lock() {
        locked.set(true)
    }

    private fun requireOwner(): LifecycleOwner = requireNotNull(owner) { "Cannot access lifecycleOwner after onDestroy." }

    private fun requireSourceObserver() = requireNotNull(sourceObserver) { "Cannot access observer after onDestroy." }

    companion object {
        private val DEFAULT_ACTIVE_STATE = State.STARTED
    }
}
